[FLINK-32695] [Tests] Add flink-test-utils-connectors module for Source V2 tests.

Poorvankbhatia opened the pull request:
## What is the purpose of the change

Added a minimal, test-only utility module for FLIP-27 (Source V2) tests to reduce boilerplate when using Source V2 in Test cases. This is being done as a part of FLINK-32695 to migrate all the legacy test sources.

While migrating legacy SourceFunction–based tests to FLIP-27 (Source V2), a recurring pattern was noticed: many tests only need a mostly no-op scaffold (e.g., a source/enumerator that does nothing) or just one component—most often the SourceReader—to exercise a specific behavior. Hence this change is being made so that those patterns can be captured and test authors can plug in a tiny override (reader/enumerator/serializer) without re-implementing boilerplate.

Module path: flink-test-utils-parent/flink-test-utils-source

## Brief change log

  - Add new module in flink-test-utils-parent (packaging: jar).
  - Introduced a reusable scaffolding under org.apache.flink.test.util.source:
    - AbstractTestSource<T> — default BOUNDED Source V2 with sensible no-op implementations; easy to extend.
    - TestSourceReader<T> — minimal SourceReader (default END_OF_INPUT), plumbs SourceReaderContext.
    - TestSplit — lightweight split with singleton-style serializer for simple tests.
    - TestSplitEnumerator — assigns a single split on request and signals no-more-splits to avoid hanging tests.
    - VoidSerializer — serializer for Void enumerator checkpoints.
  - Keep dependencies lean:
    - flink-core
  - No production modules changed.  


## Verifying this change

  - Module builds with the existing Flink parent
  - Unit tests added in SourcePatternExamplesTest exercise the new test-utils and cover complex Source V2 behaviors

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable, Javadoc added to new classes to describe intended usage and extension points.

fapaul commented on flink-test-utils-parent/flink-test-utils-source/src/main/java/org/apache/flink/test/util/source/AbstractTestSource.java:
```diff
@@ -0,0 +1,41 @@
+ *
+ * @param <T> The type of records produced by this source
+ */
+public abstract class AbstractTestSource<T> extends AbstractTestSourceBase<T, Void> {
```
That makes sense. For all use cases without the need for the enumerator checkpoints, we use the `AbstractTestSource`; otherwise, we need to fallback to implementing the base class.

Poorvankbhatia commented:
> @Poorvankbhatia can you rebase one last time on latest master and update the branch?
> 
> Unfortunately, there seems to be an issue with the CI at the moment and I cannot retrigger it. Currently, both builds on your fork and on the Azure CI are red due to "unrelated" failures, but the build stopped before executing your new tests. I ran them locally, and they pass, but it would be good to at least have one green run either in your fork or on the official CI.

Rebased with the latest master and updated the commit message. @fapaul have a look. Thanks. :)
